main/src/com/google/refine/browsing/facets/TextListFacet.java

```diff
@@ -0,0 +1,172 @@
+/*
+
+Copyright 2010, Google Inc.
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions are
+met:
+
+    * Redistributions of source code must retain the above copyright
+notice, this list of conditions and the following disclaimer.
+    * Redistributions in binary form must reproduce the above
+copyright notice, this list of conditions and the following disclaimer
+in the documentation and/or other materials provided with the
+distribution.
+    * Neither the name of Google Inc. nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,           
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY           
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+*/
+
+package com.google.refine.browsing.facets;
+
+import java.util.LinkedList;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonIgnore;
+import com.fasterxml.jackson.annotation.JsonInclude;
+import com.fasterxml.jackson.annotation.JsonInclude.Include;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.refine.ProjectManager;
+import com.google.refine.browsing.DecoratedValue;
+import com.google.refine.browsing.FilteredRecords;
+import com.google.refine.browsing.FilteredRows;
+import com.google.refine.browsing.RecordFilter;
+import com.google.refine.browsing.RowFilter;
+import com.google.refine.browsing.filters.AllRowsRecordFilter;
+import com.google.refine.browsing.filters.AnyRowRecordFilter;
+import com.google.refine.browsing.filters.ExpressionEqualRowFilter;
+import com.google.refine.browsing.util.ExpressionNominalValueGrouper;
+import com.google.refine.expr.Evaluable;
+import com.google.refine.expr.MetaParser;
+import com.google.refine.expr.ParsingException;
+import com.google.refine.model.Column;
+import com.google.refine.model.Project;
+
+public class TextListFacet extends ListFacet {
```

Comment:
I am not sure it is a good idea to make TextListFacet extend ListFacet. In the current situation, ListFacet is not designed to be extended in this way, which causes the current NullPointerExceptions in the tests.

If the desired end state is to completely replace the original ListFacet with the new TextListFacet, then I think it would be cleaner to actually duplicate the existing code, with the understanding that on the long term the duplication will disappear when we drop ListFacet.

Reply:
> I am not sure it is a good idea to make TextListFacet extend ListFacet. In the current situation, ListFacet is not designed to be extended in this way, which causes the current NullPointerExceptions in the tests.

As I noted in #2018:
>My original idea was to have an abstract class for a 'list facet' which would support grouping by any of the supported data types then be implemented by different types of list facet including the existing 'list facet' and a new 'text list facet' which would exhibit the new desired behaviour. But each time I look at this I start to think the whole way facets are structured feels not quite right, and end up fluctuating between just doing something v simple but duplicating loads of code, and revisiting the whole structure of the facet code (at which point I have to lie down in a dark room for half an hour).

When I step away from the problem, it feels like having a general 'list facet' class which then could be specialised into types seems like a good idea. Especially when considering the point of this work is to end up with the old 'list facet' behaviour alongside the new 'text list facet' behaviour. However, when I try to implement I hit the problems you note.

I also had in mind this might make it easier to support different types of behaviour in different types of list facet going forward (since discussions in #1957 and #1662 show this is difficult to agree on)

I can't escape the feeling that the code around facets (of all types) could be more efficient and easier to extend, but I suspect I don't have the experience & knowledge to implement this, and I definitely haven't had the time

Reply:
@ostephens I'd rather you not give up on this, given time this is a great opportunity to learn more!  For me being on the sidelines, and looking at the code and Antonin's comments I have learned more...and learned why I pulled away from Java in the first place long ago ;-)  

Perhaps a bit of guidance on a call with Antonin in a paired session for 30 mins - 1 hour could help OpenRefine and you?  I'll leave it to both of you to decide.  If you strongly feel like you want to pull away from this particular refactoring, then perhaps @wetneb can step in to finish this up?

Reply:
Hi @ostephens, 
> My original idea was to have an abstract class for a 'list facet' which would support grouping by any of the supported data types then be implemented by different types of list facet including the existing 'list facet' and a new 'text list facet' which would exhibit the new desired behaviour.

I've just played a bit with the facets in order to implement the Minimize/Maximize functionality. And, I had to copy the EXACT same code in the 5 Facets JS object because they don't have a generic parent. I'd say I'm kind of compelled to create a common parent, let's say `GenericFacet` that would have all the common functionality of a facet.

So, in this situation, it would have a some benefits for this issue. 1) the ListFacet object would now be ridden of all the common functionality of a facet, so you would have less code to manage and refactorize for this TextListFacet.

Also, it would test all the logic of using a parent class in the context of facets. It would be kind of a practice.

And, last added benefit, in the future, such a common parent would surely help implementing things like saving facets or else… So, would you like to work on this with me?

Regards,
  Antoine